File: LeetCode/LeetCode Practice Problems/797- All Paths From Source to Target.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void dfs(vector<vector<int>> a,vector<int> p,int u)
    {
        if(u == a.size()-1)
        {
            ans.push_back(p);
            return;
        }
        
        for(auto i : a[u])
        {
            p.push_back(i);
            dfs(a,p,i);
            p.pop_back();
        }
    }
                          
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        dfs(graph,{0},0);
        return ans;
    }
};

File: LeetCode/LeetCode Practice Problems/797-All Paths From Source to Target.java
class Solution {
    
    List<List<Integer>> ans = new ArrayList();
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<Integer> path = new ArrayList();
        
        dfs(graph , 0, path);
        
        return ans;
    }
    
    public void dfs(int[][] graph,int src,List<Integer> path) {
        path.add(src);
        
        if(src == graph.length - 1) {
            ans.add(new ArrayList<Integer>(path));
            return;
        }
        
        for(int next: graph[src]) {
            dfs(graph,next,path);
            path.remove(path.size() - 1);
        }
    }
}
